### ISSUE TITLE
libffmpeg.dylib is necessary when spawn chromedriver on macos, i think should save libffmpeg.dylib when install-chromedriver

### ISSUE THREAD

**Issue by yanguoyu** [SeleniumHQ/selenium-ide#1302]
## 🐛 Bug Report

i found it can't ```executeScript 'document.getElementById("aut").src = "https://google.com"'``` when i run yarn start on the branch trunk.

## To Reproduce

yarn
yarn build
yarn start

## Expected behavior

the aut element src should be ```https://google.com```

## Project file reproducing this issue (highly encouraged)

Please provide a project file `.side` that reproduces this issue.

Issues without a reproduction project are likely to stall.

## Environment

OS: Mac OS 11.1
Branch: trunk

**Comment by toddtarsi** [comment]
@yanguoyu - Whoa, you're cloning off trunk? That's like the incomplete v4 version that we're re-writing to run on Electron / NodeJS. That way we can run drivers and save entire directories with reckless impunity. Are you looking to help with v4? 🤩
If so, I have a branch where I was starting to do some stuff, but I kinda don't like the way I did it. I'd be absolutely honored to try and get something going with you there.

If you're looking to make smaller changes to the ide, and work on v3 (which I would also love 😍), you'll want to change to the v3 branch.

Either way, I'm excited that you're cloning and building. Please let me know if I can help further.

**Comment by yanguoyu** [comment]
yeah, i'm glad to head that. I'm honored to participate v4.What can i do something for v4?

**Comment by toddtarsi** [comment]
That is a good question. There are a few main parts to what we need for v4. To try and keep this focused, I think this is the absolute minimum total work needed:

1. We need to be able to download and run chromedriver instances. Using chromedriver, we need to start up a window and "record" it via CDP and drive it via webdriver. I believe the downloading and running functionality is somewhat there. The entire CDP side is genuinely something I don't understand at all yet. 

2. We need to take the frontend from v3, and effectively "lift and shift" it into v4. This is something I've tried a couple ways, but I think the easiest way is to run a small node server inside our electron app, and serve the v3 extension html from there. We'll then replace the privileged extension calls with server calls that will operate against our Chromedriver instance. So our save commands will call an endpoint like api/v4/save or something, export is api/v4/export, etc.

Once we have these two components functioning, we should have the main body of work done.
I can help with either one of these, but the less painful part (initially) will be lifting and shifting the UX.

How about this: Will you give me a week or two to do a bit of aligning between these two different PRs:

https://github.com/SeleniumHQ/selenium-ide/pull/1258
https://github.com/SeleniumHQ/selenium-ide/pull/1243

These were two different PRs where I went about this stuff the wrong way. The first attempt, I brought all of the code as-is and tried polyfilling the web extension API. While initially nice, this became extremely difficult around the sendMessage API, and would make things more complicated going forward. The second attempt involved re-writing the code from scratch. This actually has been nice, but I find that using the "Electron" data strategy (ipcMain and ipcRenderer) might be a terrible decision when technologies like XHR and websockets already exist. It adds a wacky layer of hand-rolled-ness to what should be standard web SPA stuff.

Basically, there is some code I like from both of these, and some code I really dislike from both of these, and I'd like to put out a bit of effort to dig through them, reconcile things a bit, and give you (and me) a base to work with that isn't setting us up for failure.

Ideally, we're able to re-use the majority of the v3 code in v4. It's a solid React SPA. And then the remaining work is just replacing the privileged extension stuff with XHR calls to our electron / node server, figuring out how best to use CDP for recording and webdriver for playback.

**Comment by yanguoyu** [comment]
I think it's a good idea to polyfill the web extension API, then we can use v3 React component.
And i want to try with [#1243](https://github.com/SeleniumHQ/selenium-ide/pull/1243).
but i need more time to understand ```webextension-polyfill```, i has't use it before.
I feel we can polyfill the web extension API by using Electron API. I will try it in the next two weeks, but i can't promise i will resolve it.

**Comment by yanguoyu** [comment]
I think it's a good idea to polyfill the web extension API, then we can use v3 React component.And i want to try with [#1243](https://github.com/SeleniumHQ/selenium-ide/pull/1243).but i need more time to understand ```webextension-polyfill```,i has't use it before.I feel we can polyfill the web extension API by using Electron API.I will try it in the next two weeks, but i can't promise i will resolve it.

**Comment by toddtarsi** [comment]
@yanguoyu - My feeling is that part of that approach is correct (re-using as much of v3 as possible is essential), but polyfilling the web extension API gets very difficult. Additionally, I think the quality of the end code would suffer from trying to convolute to maintain these boundaries.

Regardless if we go this way or another, I know one thing we could start with that would really help here:

Can you start by going through the selenium-ide package on the v3 branch, and identify all relevant places where the webextension API is consumed by selenium IDE, and move those to a single location, like a folder called 'extensionBoundary' or similar. This way, when we go to migrate by re-writing the webextension-polyfill, or migrate by replacing these calls with specialized requests to a NodeJS server (my preference), it will be much more easily understood and reasoned through. 

I think this would do a world of good. It would get you into the codebase, align us for transporting as much of v3 as possible, and help us both understand how much work this could might require.

Please let me know how this sounds. For myself personally, my next goal is to honestly just read the trunk code closer. There were probably hundreds of hours written towards this by an insanely talented individual (@corevo), and I don't know what any of the packages besides selenium-ide really do.

**Comment by toddtarsi** [comment]
@yanguoyu - Do you use Slack by the way? Is there any chance I could invite you to join the Selenium organization on Slack? This way you can DM me any time :)

https://www.selenium.dev/support/

**Comment by yanguoyu** [comment]
@toddtarsi of cause, i has  join the Selenium organization on slack, my name is guoyu

**Comment by toddtarsi** [comment]
I'm going to close this issue for now. The binaries work well and this builds plenty fine on a variety of platforms. Thanks for your help @yanguoyu
